component/src/governance/check.rs

```diff
@@ -291,4 +261,81 @@ pub mod stateful {
 
         Ok(())
     }
+
+    pub async fn proposal_withdraw(
+        state: &State,
+        proposal_withdraw: &ProposalWithdraw,
+    ) -> Result<()> {
+        // Any voteable proposal can be withdrawn
+        proposal_voteable(state, proposal_withdraw.proposal).await?;
+        Ok(())
+    }
+
+    pub async fn proposal_deposit_claim(
+        state: &State,
+        proposal_deposit_claim: &ProposalDepositClaim,
+    ) -> Result<()> {
+        // Any finished proposal can have its deposit claimed
+        proposal_claimable(state, proposal_deposit_claim.proposal).await?;
+        // Check that the deposit amount matches the proposal being claimed
+        proposal_claim_valid_deposit(
+            state,
+            proposal_deposit_claim.proposal,
+            proposal_deposit_claim.deposit_amount,
+        )
+        .await?;
+        Ok(())
+    }
+
+    async fn proposal_claimable(state: &State, proposal_id: u64) -> Result<()> {
+        if let Some(proposal_state) = state.proposal_state(proposal_id).await? {
+            use proposal::State::*;
+            match proposal_state {
+                Voting => {
+                    anyhow::bail!("proposal {} is still voting", proposal_id)
+                }
+                Withdrawn { .. } => {
+                    anyhow::bail!(
+                        "proposal {} has been withdrawn but voting has not concluded",
+                        proposal_id
+                    )
+                }
+                Finished { .. } => {
+                    // This is when you can claim a proposal
+                }
```

Comment:
The only success case is when the proposal is precisely in the finished-but-not-claimed state. It transitions out of this when you claim it.